# NovaMart Customer Retention and Campaign Optimization
## Applied Analytical Models — Final Course Project

This notebook is organized according to the seven required project tasks.

**Current progress:** Tasks 1 and 2 are complete. Tasks 3–7 remain intentionally unstarted.

# Task 1 — Data Understanding and Attribute Classification

### Objective
Understand what each dataset represents, identify the data object, dataset sizes, target variable, identifier, candidate predictors, and classify every attribute by analytical type and measurement scale.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## 1.1 Load the raw datasets

The historical dataset contains past customers whose churn outcome is known.  
The current dataset contains customers whose churn risk must be predicted later.

In [2]:
historical = pd.read_csv("NovaMart_Historical_Customers.csv")
current = pd.read_csv("NovaMart_Current_Customers.csv")

print("Datasets loaded successfully.")

Datasets loaded successfully.


## 1.2 Confirm dataset dimensions

In [3]:
print("Historical dataset shape:", historical.shape)
print("Current dataset shape:", current.shape)

Historical dataset shape: (1215, 20)
Current dataset shape: (300, 19)


**Confirmed record definition and dimensions:** One record (row) represents one NovaMart customer observed at the customer level. The historical dataset contains **1,215 customers and 20 variables**; the current dataset contains **300 customers and 19 variables**. The one-column difference is the historical outcome `Churn`.

## 1.3 Preview the datasets

In [4]:
print("Historical customers:")
display(historical.head())

print("\nCurrent customers:")
display(current.head())

Historical customers:

Current customers:


,Customer_ID,Age,Gender,Region,Membership_Level,Tenure_Months,Monthly_Income_JD,Monthly_Spend_JD,Total_Orders,Average_Order_Value_JD,Days_Since_Last_Purchase,Website_Visits_Last_30_Days,Email_Engagement_Rate,Discount_Usage_Rate,Satisfaction_Score,Support_Tickets_Last_6_Months,Complaint_Flag,Last_Campaign_Response,Payment_Method,Churn
0,C00370,60.0,Male,Balqa,Basic,13,1810.15,270.10,12,306.12,10,24,0.393,0.525,5.0,2,0,0,Digital Wallet,0
1,C01187,67.0,Male,Amman,Basic,11,2175.33,87.14,8,109.86,9,18,0.461,0.464,4.0,1,0,0,Digital Wallet,0
2,C00806,50.0,Female,Amman,Silver,48,2087.84,201.90,23,365.24,11,12,0.053,0.083,4.0,0,0,0,Card,0
3,C00953,42.0,Female,Zarqa,Platinum,41,2079.61,233.06,19,490.29,75,13,0.414,0.133,3.0,1,0,1,Cash,0
4,C00236,27.0,Female,Irbid,Platinum,5,1912.70,300.74,7,182.06,67,15,0.297,0.795,5.0,1,0,0,Cash,0


,Customer_ID,Age,Gender,Region,Membership_Level,Tenure_Months,Monthly_Income_JD,Monthly_Spend_JD,Total_Orders,Average_Order_Value_JD,Days_Since_Last_Purchase,Website_Visits_Last_30_Days,Email_Engagement_Rate,Discount_Usage_Rate,Satisfaction_Score,Support_Tickets_Last_6_Months,Complaint_Flag,Last_Campaign_Response,Payment_Method
0,C20001,24.0,Female,Irbid,Platinum,84,2074.52,264.83,37,500.00,4,21,0.183,0.914,4.0,0,0,0,Card
1,C20002,44.0,Male,Amman,Gold,12,2647.38,222.34,8,356.05,70,21,0.371,0.576,4.0,0,0,0,Bank Transfer
2,C20003,39.0,Female,Zarqa,Gold,12,1296.71,215.81,11,264.25,9,23,0.543,0.748,4.0,2,0,1,digital wallet
3,C20004,62.0,Female,Zarqa,Silver,84,2577.75,191.53,39,438.68,72,22,0.432,0.688,3.0,2,0,1,Card
4,C20005,-4.0,Male,Irbid,Basic,19,1429.96,90.99,5,335.50,28,18,0.098,0.518,4.0,2,0,0,Card


## 1.4 Inspect column names and data types

In [5]:
print("Historical columns:")
print(historical.columns.tolist())

print("\nCurrent columns:")
print(current.columns.tolist())

Historical columns:
['Customer_ID', 'Age', 'Gender', 'Region', 'Membership_Level', 'Tenure_Months', 'Monthly_Income_JD', 'Monthly_Spend_JD', 'Total_Orders', 'Average_Order_Value_JD', 'Days_Since_Last_Purchase', 'Website_Visits_Last_30_Days', 'Email_Engagement_Rate', 'Discount_Usage_Rate', 'Satisfaction_Score', 'Support_Tickets_Last_6_Months', 'Complaint_Flag', 'Last_Campaign_Response', 'Payment_Method', 'Churn']

Current columns:
['Customer_ID', 'Age', 'Gender', 'Region', 'Membership_Level', 'Tenure_Months', 'Monthly_Income_JD', 'Monthly_Spend_JD', 'Total_Orders', 'Average_Order_Value_JD', 'Days_Since_Last_Purchase', 'Website_Visits_Last_30_Days', 'Email_Engagement_Rate', 'Discount_Usage_Rate', 'Satisfaction_Score', 'Support_Tickets_Last_6_Months', 'Complaint_Flag', 'Last_Campaign_Response', 'Payment_Method']


In [6]:
print("Historical pandas data types:")
display(historical.dtypes.rename("Pandas_Dtype").to_frame())

print("Current pandas data types:")
display(current.dtypes.rename("Pandas_Dtype").to_frame())

print("Detailed historical dataset information:")
historical.info()

print("\nDetailed current dataset information:")
current.info()

Historical pandas data types:
Current pandas data types:
Detailed historical dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 1215 entries, 0 to 1214
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Customer_ID                    1215 non-null   str    
 1   Age                            1189 non-null   float64
 2   Gender                         1215 non-null   str    
 3   Region                         1198 non-null   str    
 4   Membership_Level               1215 non-null   str    
 5   Tenure_Months                  1215 non-null   int64  
 6   Monthly_Income_JD              1193 non-null   float64
 7   Monthly_Spend_JD               1194 non-null   float64
 8   Total_Orders                   1215 non-null   int64  
 9   Average_Order_Value_JD         1215 non-null   float64
 10  Days_Since_Last_Purchase       1215 non-null   int64  
 11  Website_Visits_Last_3

,Pandas_Dtype
Customer_ID,str
Age,float64
Gender,str
Region,str
Membership_Level,str
Tenure_Months,int64
Monthly_Income_JD,float64
Monthly_Spend_JD,float64
Total_Orders,int64
Average_Order_Value_JD,float64


,Pandas_Dtype
Customer_ID,str
Age,float64
Gender,str
Region,str
Membership_Level,str
Tenure_Months,int64
Monthly_Income_JD,float64
Monthly_Spend_JD,float64
Total_Orders,int64
Average_Order_Value_JD,float64


**Note:** Pandas data types such as `str`, `int64`, and `float64` describe how values are stored in Python.  
They are not the same as statistical measurement types such as nominal, ordinal, binary, or ratio.

## 1.5 Identify the target, identifier, and predictor variables

In [7]:
target = "Churn"
identifier = "Customer_ID"

predictors = [
    col for col in historical.columns
    if col not in [target, identifier]
]

# Executable confirmations of the Task 1 interpretation.
assert historical.shape == (1215, 20)
assert current.shape == (300, 19)
assert identifier in historical.columns and identifier in current.columns
assert historical[identifier].notna().all() and current[identifier].notna().all()
assert target in historical.columns and target not in current.columns
assert set(historical[target].dropna().unique()) == {0, 1}
assert predictors == [col for col in current.columns if col != identifier]

print("Record represented by one row: one NovaMart customer")
print("Identifier:", identifier, "(intended customer identifier)")
print("Historical distinct IDs:", historical[identifier].nunique(), "of", len(historical), "rows")
print("Historical duplicate-ID occurrences:", historical[identifier].duplicated().sum())
print("Current distinct IDs:", current[identifier].nunique(), "of", len(current), "rows")
print("Target variable:", target, "(historical only; coded 0/1)")
print("Number of candidate predictor variables:", len(predictors))
print("\nPredictor variables:")
for col in predictors:
    print("-", col)
print("\nAll Task 1 structural checks passed.")

Record represented by one row: one NovaMart customer
Identifier: Customer_ID (intended customer identifier)
Historical distinct IDs: 1200 of 1215 rows
Historical duplicate-ID occurrences: 15
Current distinct IDs: 300 of 300 rows
Target variable: Churn (historical only; coded 0/1)
Number of candidate predictor variables: 18

Predictor variables:
- Age
- Gender
- Region
- Membership_Level
- Tenure_Months
- Monthly_Income_JD
- Monthly_Spend_JD
- Total_Orders
- Average_Order_Value_JD
- Days_Since_Last_Purchase
- Website_Visits_Last_30_Days
- Email_Engagement_Rate
- Discount_Usage_Rate
- Satisfaction_Score
- Support_Tickets_Last_6_Months
- Complaint_Flag
- Last_Campaign_Response
- Payment_Method

All Task 1 structural checks passed.


`Customer_ID` is the intended identifier in both datasets: it labels a customer and is excluded from modelling because its codes do not measure customer characteristics. The current IDs are unique, but the historical file has 1,200 distinct IDs in 1,215 rows (15 duplicate-ID occurrences). This is recorded as a data-quality issue for Task 2; no rows are altered here. `Churn` is the historical binary target (`0` = did not churn, `1` = churned). The other 18 variables shared by both datasets are candidate predictors.


## 1.6 Inspect the churn target distribution

In [8]:
churn_counts = historical["Churn"].value_counts().sort_index()
churn_percentages = historical["Churn"].value_counts(normalize=True).sort_index() * 100

churn_summary = pd.DataFrame({
    "Count": churn_counts,
    "Percentage": churn_percentages.round(2)
})

churn_summary.index.name = "Churn"
display(churn_summary)

,Count,Percentage
Churn,,
0,961,79.09
1,254,20.91


**Interpretation:** `Churn = 0` represents a customer who did not churn, while `Churn = 1` represents a customer who churned.  
The distribution is useful later when evaluating whether class imbalance needs to be considered during predictive modelling.

## 1.7 Attribute classification table

In [9]:
classification_rows = [
    # Attribute, Role, Variable nature, Measurement scale, Numeric form, Concise meaning
    ("Customer_ID", "Identifier", "Qualitative", "Nominal", "Not applicable", "Intended customer identifier"),
    ("Age", "Predictor", "Quantitative", "Ratio", "Continuous (recorded in years)", "Customer age"),
    ("Gender", "Predictor", "Qualitative", "Nominal", "Not applicable", "Customer gender category"),
    ("Region", "Predictor", "Qualitative", "Nominal", "Not applicable", "Customer region"),
    ("Membership_Level", "Predictor", "Qualitative", "Ordinal", "Not applicable", "Ordered loyalty tier"),
    ("Tenure_Months", "Predictor", "Quantitative", "Ratio", "Integer", "Months as a customer"),
    ("Monthly_Income_JD", "Predictor", "Quantitative", "Ratio", "Continuous", "Monthly income in JD"),
    ("Monthly_Spend_JD", "Predictor", "Quantitative", "Ratio", "Continuous", "Monthly NovaMart spend in JD"),
    ("Total_Orders", "Predictor", "Quantitative", "Ratio", "Integer", "Cumulative order count"),
    ("Average_Order_Value_JD", "Predictor", "Quantitative", "Ratio", "Continuous", "Average order value in JD"),
    ("Days_Since_Last_Purchase", "Predictor", "Quantitative", "Ratio", "Integer", "Purchase recency in days"),
    ("Website_Visits_Last_30_Days", "Predictor", "Quantitative", "Ratio", "Integer", "Website visit count in 30 days"),
    ("Email_Engagement_Rate", "Predictor", "Quantitative", "Ratio", "Continuous", "Proportion of email engagement"),
    ("Discount_Usage_Rate", "Predictor", "Quantitative", "Ratio", "Continuous", "Proportion of purchases using discounts"),
    ("Satisfaction_Score", "Predictor", "Qualitative", "Ordinal", "Integer-coded category", "Ordered satisfaction rating"),
    ("Support_Tickets_Last_6_Months", "Predictor", "Quantitative", "Ratio", "Integer", "Support ticket count in six months"),
    ("Complaint_Flag", "Predictor", "Qualitative", "Binary", "Integer-coded category", "Whether a complaint was recorded"),
    ("Last_Campaign_Response", "Predictor", "Qualitative", "Binary", "Integer-coded category", "Whether the last campaign received a response"),
    ("Payment_Method", "Predictor", "Qualitative", "Nominal", "Not applicable", "Payment method category"),
    ("Churn", "Target", "Qualitative", "Binary", "Integer-coded category", "Whether the customer churned"),
]

attribute_classification = pd.DataFrame(
    classification_rows,
    columns=[
        "Attribute", "Role", "Variable_Type", "Measurement_Type",
        "Numeric_Form", "Meaning"
    ],
)
attribute_classification.insert(
    2,
    "Pandas_Dtype",
    attribute_classification["Attribute"].map(
        historical.dtypes.astype(str).to_dict()
    ),
)

assert attribute_classification["Attribute"].tolist() == historical.columns.tolist()
assert attribute_classification["Attribute"].is_unique

display(attribute_classification)
print(f"Classified all {len(attribute_classification)} historical variables.")

Classified all 20 historical variables.


,Attribute,Role,Pandas_Dtype,Variable_Type,Measurement_Type,Numeric_Form,Meaning
0,Customer_ID,Identifier,str,Qualitative,Nominal,Not applicable,Intended customer identifier
1,Age,Predictor,float64,Quantitative,Ratio,Continuous (recorded in years),Customer age
2,Gender,Predictor,str,Qualitative,Nominal,Not applicable,Customer gender category
3,Region,Predictor,str,Qualitative,Nominal,Not applicable,Customer region
4,Membership_Level,Predictor,str,Qualitative,Ordinal,Not applicable,Ordered loyalty tier
5,Tenure_Months,Predictor,int64,Quantitative,Ratio,Integer,Months as a customer
6,Monthly_Income_JD,Predictor,float64,Quantitative,Ratio,Continuous,Monthly income in JD
7,Monthly_Spend_JD,Predictor,float64,Quantitative,Ratio,Continuous,Monthly NovaMart spend in JD
8,Total_Orders,Predictor,int64,Quantitative,Ratio,Integer,Cumulative order count
9,Average_Order_Value_JD,Predictor,float64,Quantitative,Ratio,Continuous,Average order value in JD


## Task 1 Conclusion

The data object is a **NovaMart customer**, so each row is one customer-level record. The historical data has **1,215 records × 20 variables**, while the current data has **300 records × 19 variables** and omits the outcome that has not yet been observed.

`Customer_ID` is the intended identifier, `Churn` is the binary target, and the 18 other shared attributes are candidate predictors. The current file has 300 unique IDs; the historical file has 1,200 distinct IDs across 1,215 rows, revealing 15 duplicate-ID occurrences to address in Task 2. In this run, pandas stores the columns as `str`, `int64`, or `float64`; these storage dtypes are shown separately from their analytical classifications.

The attribute table classifies every historical variable by role, qualitative/quantitative nature, measurement type, and numeric form. Nominal values are unordered labels, binary values have two categories, ordinal values have a meaningful order, and ratio values are numeric measures with a meaningful zero. **No variable is classified as interval scale** because none has meaningful differences paired with an arbitrary zero. “Integer” is recorded separately as a numeric form because a count can be both integer-valued and ratio-scaled.

The target contains **961 non-churn customers (79.09%)** and **254 churn customers (20.91%)**. This imbalance is noted for later modelling, but no Task 2 preparation or modelling is performed here.

# Task 2 — Data Preparation and Quality

### Objective
Assess both raw datasets for accuracy, completeness, consistency, validity, uniqueness, and timeliness before making any change. The assessment is displayed first; cleaning is then applied through explicit, compatible rules. The raw CSV files are read-only inputs.

## 2.1 Systematic pre-cleaning assessment

The checks below report missing values, exact duplicate rows, duplicate identifiers, category spellings, domain validity, pandas dtypes, and Tukey IQR outliers. An IQR flag is a diagnostic—not automatic proof that a value is wrong.

In [10]:
from pathlib import Path

historical_raw = historical.copy(deep=True)
current_raw = current.copy(deep=True)
datasets_raw = {"Historical": historical_raw, "Current": current_raw}

categorical_columns = ["Gender", "Region", "Membership_Level", "Payment_Method"]
numeric_columns = [
    "Age", "Tenure_Months", "Monthly_Income_JD", "Monthly_Spend_JD",
    "Total_Orders", "Average_Order_Value_JD", "Days_Since_Last_Purchase",
    "Website_Visits_Last_30_Days", "Email_Engagement_Rate",
    "Discount_Usage_Rate", "Satisfaction_Score",
    "Support_Tickets_Last_6_Months"
]

quality_rows = []
for dataset_name, df in datasets_raw.items():
    for column, count in df.isna().sum().items():
        if count:
            quality_rows.append([dataset_name, "Completeness", f"Missing: {column}", int(count), "Requires treatment"])

    quality_rows.append([dataset_name, "Uniqueness", "Exact duplicate rows", int(df.duplicated().sum()), "Remove repeated copies only"])
    quality_rows.append([dataset_name, "Uniqueness", "Duplicate Customer_ID occurrences", int(df["Customer_ID"].duplicated().sum()), "Inspect against exact duplicates"])

    for column in categorical_columns:
        trimmed_title = df[column].astype("string").str.strip().str.title()
        variants = int((df[column].astype("string") != trimmed_title).fillna(False).sum())
        quality_rows.append([dataset_name, "Consistency", f"Non-canonical labels: {column}", variants, "Standardize case/spacing"])

    invalid_rules = {
        "Age outside 18–100": (~df["Age"].between(18, 100) & df["Age"].notna()),
        "Negative Monthly_Income_JD": df["Monthly_Income_JD"] < 0,
        "Negative Monthly_Spend_JD": df["Monthly_Spend_JD"] < 0,
        "Satisfaction_Score outside integer 1–5": (((~df["Satisfaction_Score"].between(1, 5)) | (df["Satisfaction_Score"] % 1 != 0)) & df["Satisfaction_Score"].notna()),
        "Invalid rate outside 0–1": ((~df["Email_Engagement_Rate"].between(0, 1) & df["Email_Engagement_Rate"].notna()) | (~df["Discount_Usage_Rate"].between(0, 1) & df["Discount_Usage_Rate"].notna())),
        "Invalid binary flag": ((~df["Complaint_Flag"].isin([0, 1])) | (~df["Last_Campaign_Response"].isin([0, 1]))),
        "Invalid Customer_ID format": ~df["Customer_ID"].astype("string").str.fullmatch(r"C\d{5}", na=False),
    }
    if "Churn" in df:
        invalid_rules["Invalid Churn code"] = ~df["Churn"].isin([0, 1])
    for issue, mask in invalid_rules.items():
        quality_rows.append([dataset_name, "Validity / accuracy", issue, int(mask.sum()), "Correct if non-zero"])

    quality_rows.append([dataset_name, "Data types", "Columns with non-expected storage type", 0, "Types are compatible; Age and Satisfaction are float because of missing values"])
    quality_rows.append([dataset_name, "Timeliness", "Observation timestamp unavailable", None, "Cannot verify dataset age; recency field is usable but is not an extract date"])

quality_assessment = pd.DataFrame(
    quality_rows,
    columns=["Dataset", "Quality_Dimension", "Check_or_Issue", "Affected_Rows_or_Values", "Decision"]
)
display(quality_assessment)

,Dataset,Quality_Dimension,Check_or_Issue,Affected_Rows_or_Values,Decision
0,Historical,Completeness,Missing: Age,26.0,Requires treatment
1,Historical,Completeness,Missing: Region,17.0,Requires treatment
2,Historical,Completeness,Missing: Monthly_Income_JD,22.0,Requires treatment
3,Historical,Completeness,Missing: Monthly_Spend_JD,21.0,Requires treatment
4,Historical,Completeness,Missing: Email_Engagement_Rate,26.0,Requires treatment
5,Historical,Completeness,Missing: Satisfaction_Score,20.0,Requires treatment
6,Historical,Uniqueness,Exact duplicate rows,15.0,Remove repeated copies only
7,Historical,Uniqueness,Duplicate Customer_ID occurrences,15.0,Inspect against exact duplicates
8,Historical,Consistency,Non-canonical labels: Gender,7.0,Standardize case/spacing
9,Historical,Consistency,Non-canonical labels: Region,21.0,Standardize case/spacing


### Assessment interpretation

- **Accuracy and validity:** age must be 18–100, monetary values cannot be negative, satisfaction must be an integer from 1 to 5, rates must be within 0–1, and flags/target must be binary. Invalid values are not overwritten during assessment.
- **Completeness:** missingness is counted column by column rather than hidden inside a single total.
- **Consistency:** capitalization and trailing spaces create duplicate category meanings.
- **Uniqueness:** duplicate rows and duplicate IDs are checked separately. In the historical file, every duplicate-ID occurrence is an exact repeated row.
- **Timeliness:** neither file contains an extraction or observation date, so source freshness cannot be confirmed. `Days_Since_Last_Purchase` measures customer recency, not dataset timeliness.

In [11]:
print("Missing values before cleaning:")
missing_before = pd.concat(
    {name: df.isna().sum() for name, df in datasets_raw.items()}, axis=1
)
display(missing_before[missing_before.sum(axis=1) > 0])

print("Duplicate diagnostics:")
duplicate_summary = pd.DataFrame({
    name: {
        "Exact duplicate rows": int(df.duplicated().sum()),
        "Duplicate-ID occurrences": int(df["Customer_ID"].duplicated().sum()),
        "All rows sharing a duplicated ID": int(df["Customer_ID"].duplicated(keep=False).sum()),
    }
    for name, df in datasets_raw.items()
}).T
display(duplicate_summary)

print("Category labels before cleaning:")
for name, df in datasets_raw.items():
    print(f"\n{name}")
    for column in categorical_columns:
        print(f"{column}: {df[column].value_counts(dropna=False).to_dict()}")

Missing values before cleaning:
Duplicate diagnostics:
Category labels before cleaning:

Historical
Gender: {'Male': 607, 'Female': 594, 'female': 7, 'M': 7}
Region: {'Amman': 561, 'Irbid': 210, 'Zarqa': 174, 'Balqa': 127, 'Other': 105, nan: 17, 'Zarqa ': 7, 'amman': 7, 'IRBID': 7}
Membership_Level: {'Basic': 468, 'Silver': 343, 'Gold': 249, 'Platinum': 134, 'GOLD': 7, 'basic': 7, 'Silver ': 7}
Payment_Method: {'Card': 545, 'Digital Wallet': 327, 'Cash': 225, 'Bank Transfer': 104, 'digital wallet': 7, 'Bank transfer': 7}

Current
Gender: {'Female': 151, 'Male': 145, 'female': 2, 'M': 2}
Region: {'Amman': 132, 'Irbid': 61, 'Zarqa': 43, 'Balqa': 31, 'Other': 23, nan: 4, 'amman': 2, 'IRBID': 2, 'Zarqa ': 2}
Membership_Level: {'Basic': 108, 'Silver': 88, 'Gold': 64, 'Platinum': 34, 'basic': 2, 'Silver ': 2, 'GOLD': 2}
Payment_Method: {'Card': 141, 'Digital Wallet': 71, 'Cash': 54, 'Bank Transfer': 30, 'digital wallet': 2, 'Bank transfer': 2}


,Historical,Current
Age,26,6.0
Region,17,4.0
Monthly_Income_JD,22,4.0
Monthly_Spend_JD,21,5.0
Email_Engagement_Rate,26,6.0
Satisfaction_Score,20,5.0


,Exact duplicate rows,Duplicate-ID occurrences,All rows sharing a duplicated ID
Historical,15,15,30
Current,0,0,0


## 2.2 Outlier assessment before cleaning

Tukey's 1.5×IQR rule identifies statistically unusual observations; it does not prove that they are erroneous. A final review also reconstructs the earlier caps for four operational attributes using the historical 3×IQR outer fence. The previous implementation used the largest historical observation inside that fence as its cap, which was more restrictive than the fence itself.

There is no supplied business rule establishing maximum income, spend, website visits, or support tickets. All reviewed values are nonnegative and technically possible. They are therefore retained as legitimate outliers rather than altered during data-quality preparation.

In [12]:
outlier_rows = []
for dataset_name, df in datasets_raw.items():
    for column in numeric_columns:
        q1, q3 = df[column].quantile([0.25, 0.75])
        iqr = q3 - q1
        lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
        mask = (df[column] < lower) | (df[column] > upper)
        if mask.any():
            outlier_rows.append([
                dataset_name, column, int(mask.sum()), round(lower, 3),
                round(upper, 3), float(df.loc[mask, column].min()),
                float(df.loc[mask, column].max())
            ])

outlier_assessment = pd.DataFrame(outlier_rows, columns=[
    "Dataset", "Attribute", "IQR_Flag_Count", "Lower_Fence",
    "Upper_Fence", "Flagged_Min", "Flagged_Max"
])
display(outlier_assessment)

review_columns = [
    "Monthly_Income_JD", "Monthly_Spend_JD",
    "Website_Visits_Last_30_Days", "Support_Tickets_Last_6_Months"
]
previous_caps = {
    "Monthly_Income_JD": 3292.65,
    "Monthly_Spend_JD": 549.57,
    "Website_Visits_Last_30_Days": 35,
    "Support_Tickets_Last_6_Months": 8,
}
review_rows = []
historical_review = historical_raw.drop_duplicates()
for column in review_columns:
    q1, q3 = historical_review[column].quantile([0.25, 0.75])
    iqr = q3 - q1
    outer_upper = q3 + 3 * iqr
    old_cap = previous_caps[column]
    for dataset_name, df in datasets_raw.items():
        unique_df = df.drop_duplicates()
        original_values = unique_df.loc[unique_df[column] > old_cap, column].sort_values().tolist()
        review_rows.append([
            dataset_name, column, round(q1, 3), round(q3, 3),
            round(iqr, 3), round(outer_upper, 3), old_cap,
            original_values, len(original_values),
            "Plausible; retain unchanged"
        ])

cap_review = pd.DataFrame(review_rows, columns=[
    "Dataset", "Attribute", "Historical_Q1", "Historical_Q3",
    "Historical_IQR", "3xIQR_Upper_Fence", "Previous_Cap",
    "Original_Values_Previously_Capped", "Values_Affected", "Final_Decision"
])
display(cap_review)

,Dataset,Attribute,IQR_Flag_Count,Lower_Fence,Upper_Fence,Flagged_Min,Flagged_Max
0,Historical,Age,1,-12.000,100.000,105.0,105.0
1,Historical,Tenure_Months,44,-21.500,78.500,80.0,84.0
2,Historical,Monthly_Income_JD,5,297.070,3248.830,-250.0,12000.0
3,Historical,Monthly_Spend_JD,28,-31.761,403.929,-85.0,1800.0
4,Historical,Total_Orders,36,-5.000,35.000,36.0,58.0
5,Historical,Days_Since_Last_Purchase,57,-58.500,129.500,131.0,600.0
6,Historical,Website_Visits_Last_30_Days,13,3.000,27.000,28.0,350.0
7,Historical,Satisfaction_Score,10,1.500,5.500,0.0,6.0
8,Historical,Support_Tickets_Last_6_Months,7,-2.000,6.000,7.0,31.0
9,Current,Age,1,-9.500,98.500,105.0,105.0


,Dataset,Attribute,Historical_Q1,Historical_Q3,Historical_IQR,3xIQR_Upper_Fence,Previous_Cap,Original_Values_Previously_Capped,Values_Affected,Final_Decision
0,Historical,Monthly_Income_JD,1402.790,2141.145,738.355,4356.210,3292.65,[12000.0],1,Plausible; retain unchanged
1,Current,Monthly_Income_JD,1402.790,2141.145,738.355,4356.210,3292.65,"[3433.17, 12000.0]",2,Plausible; retain unchanged
2,Historical,Monthly_Spend_JD,132.365,241.575,109.210,569.205,549.57,"[610.24, 723.71, 1800.0]",3,Plausible; retain unchanged
3,Current,Monthly_Spend_JD,132.365,241.575,109.210,569.205,549.57,"[559.3, 1800.0]",2,Plausible; retain unchanged
4,Historical,Website_Visits_Last_30_Days,12.000,18.000,6.000,36.000,35.00,"[220, 350]",2,Plausible; retain unchanged
5,Current,Website_Visits_Last_30_Days,12.000,18.000,6.000,36.000,35.00,"[220, 350]",2,Plausible; retain unchanged
6,Historical,Support_Tickets_Last_6_Months,1.000,3.000,2.000,9.000,8.00,"[25, 31]",2,Plausible; retain unchanged
7,Current,Support_Tickets_Last_6_Months,1.000,3.000,2.000,9.000,8.00,"[25, 31]",2,Plausible; retain unchanged


## 2.3 Cleaning decisions

The following sequence is explicit and reproducible:

1. Remove exact duplicate rows only. No distinct customer record is discarded.
2. Strip whitespace and map known case variants to canonical category labels.
3. Convert impossible age, negative income/spend, and invalid satisfaction values to missing values.
4. Retain all nonnegative outliers in income, spending, website visits, and support tickets. Statistical rarity alone is not a defensible reason to change a valid customer observation.
5. Impute numeric missing values with historical medians and categorical missing values with historical modes. The same historical reference values are applied to the current data.
6. Restore conceptually integer-valued `Age` and `Satisfaction_Score` after imputation. No features are encoded, standardized, capped, or removed.

In [13]:
clean_historical = historical_raw.drop_duplicates().copy()
clean_current = current_raw.drop_duplicates().copy()
datasets_clean = {"Historical": clean_historical, "Current": clean_current}

cleaning_log = []
for name, raw_df in datasets_raw.items():
    cleaning_log.append([name, "Remove exact duplicate rows", int(raw_df.duplicated().sum())])

category_maps = {
    "Gender": {"M": "Male", "Male": "Male", "Female": "Female", "female": "Female"},
    "Region": {"Amman": "Amman", "amman": "Amman", "Irbid": "Irbid", "IRBID": "Irbid", "Zarqa": "Zarqa", "Balqa": "Balqa", "Other": "Other"},
    "Membership_Level": {"Basic": "Basic", "basic": "Basic", "Silver": "Silver", "Gold": "Gold", "GOLD": "Gold", "Platinum": "Platinum"},
    "Payment_Method": {"Card": "Card", "Cash": "Cash", "Digital Wallet": "Digital Wallet", "digital wallet": "Digital Wallet", "Bank Transfer": "Bank Transfer", "Bank transfer": "Bank Transfer"},
}

for name, df in datasets_clean.items():
    for column, mapping in category_maps.items():
        before = df[column].copy()
        stripped = df[column].astype("string").str.strip()
        df[column] = stripped.map(mapping).astype("string")
        changed = int((before.astype("string") != df[column]).fillna(False).sum())
        cleaning_log.append([name, f"Canonicalize {column}", changed])

    invalid_masks = {
        "Age outside 18–100 → missing": (~df["Age"].between(18, 100) & df["Age"].notna()),
        "Negative Monthly_Income_JD → missing": df["Monthly_Income_JD"] < 0,
        "Negative Monthly_Spend_JD → missing": df["Monthly_Spend_JD"] < 0,
        "Invalid Satisfaction_Score → missing": (((~df["Satisfaction_Score"].between(1, 5)) | (df["Satisfaction_Score"] % 1 != 0)) & df["Satisfaction_Score"].notna()),
    }
    invalid_columns = {
        "Age outside 18–100 → missing": "Age",
        "Negative Monthly_Income_JD → missing": "Monthly_Income_JD",
        "Negative Monthly_Spend_JD → missing": "Monthly_Spend_JD",
        "Invalid Satisfaction_Score → missing": "Satisfaction_Score",
    }
    for rule, mask in invalid_masks.items():
        cleaning_log.append([name, rule, int(mask.sum())])
        df.loc[mask, invalid_columns[rule]] = np.nan

# Final outlier decision: no capping is performed.
# The reviewed observations are statistically unusual but not invalid.

display(pd.DataFrame(cleaning_log, columns=["Dataset", "Cleaning_Rule", "Affected_Rows_or_Values"]))

,Dataset,Cleaning_Rule,Affected_Rows_or_Values
0,Historical,Remove exact duplicate rows,15
1,Current,Remove exact duplicate rows,0
2,Historical,Canonicalize Gender,14
3,Historical,Canonicalize Region,21
4,Historical,Canonicalize Membership_Level,21
5,Historical,Canonicalize Payment_Method,14
6,Historical,Age outside 18–100 → missing,3
7,Historical,Negative Monthly_Income_JD → missing,1
8,Historical,Negative Monthly_Spend_JD → missing,2
9,Historical,Invalid Satisfaction_Score → missing,2


## 2.4 Missing-value treatment and compatible data types

Median imputation is robust to skew and the retained legitimate outliers. Region uses the historical mode. The historical reference statistics are reused for current customers so later scoring receives the same preparation logic. This is dataset preparation only—not feature scaling, capping, or model fitting.

In [14]:
imputation_values = {
    column: clean_historical[column].median()
    for column in numeric_columns
}
imputation_values["Region"] = clean_historical["Region"].mode(dropna=True).iat[0]

imputation_log = []
for name, df in datasets_clean.items():
    for column, value in imputation_values.items():
        count = int(df[column].isna().sum())
        if count:
            df[column] = df[column].fillna(value)
            imputation_log.append([name, column, count, value])

    df["Age"] = df["Age"].round().astype("int64")
    df["Satisfaction_Score"] = df["Satisfaction_Score"].round().astype("int64")

display(pd.DataFrame(imputation_log, columns=["Dataset", "Attribute", "Values_Imputed", "Historical_Reference_Value"]))

,Dataset,Attribute,Values_Imputed,Historical_Reference_Value
0,Historical,Age,29,44.0
1,Historical,Monthly_Income_JD,22,1792.625
2,Historical,Monthly_Spend_JD,23,180.52
3,Historical,Email_Engagement_Rate,26,0.4165
4,Historical,Satisfaction_Score,22,4.0
5,Historical,Region,17,Amman
6,Current,Age,9,44.0
7,Current,Monthly_Income_JD,5,1792.625
8,Current,Monthly_Spend_JD,7,180.52
9,Current,Email_Engagement_Rate,6,0.4165


## 2.5 Post-cleaning validation and export

The final checks require no missing values, exact duplicates, duplicate IDs, invalid domains, or inconsistent schemas. Historical retains `Churn`; current retains the same predictor columns and identifier. The clean files are exported to `outputs/`, leaving both raw CSV files unchanged.

In [15]:
assert clean_historical.shape == (1200, 20)
assert clean_current.shape == (300, 19)
assert clean_historical.isna().sum().sum() == 0
assert clean_current.isna().sum().sum() == 0
assert clean_historical.duplicated().sum() == 0
assert clean_current.duplicated().sum() == 0
assert clean_historical["Customer_ID"].is_unique
assert clean_current["Customer_ID"].is_unique
assert clean_historical["Age"].between(18, 100).all() and clean_current["Age"].between(18, 100).all()
assert clean_historical["Satisfaction_Score"].between(1, 5).all() and clean_current["Satisfaction_Score"].between(1, 5).all()
assert (clean_historical[["Monthly_Income_JD", "Monthly_Spend_JD"]] >= 0).all().all()
assert (clean_current[["Monthly_Income_JD", "Monthly_Spend_JD"]] >= 0).all().all()
assert clean_historical["Churn"].isin([0, 1]).all()
assert clean_historical.drop(columns="Churn").columns.tolist() == clean_current.columns.tolist()

post_clean_summary = pd.DataFrame({
    "Historical": [len(clean_historical), clean_historical.shape[1], int(clean_historical.isna().sum().sum()), int(clean_historical.duplicated().sum()), int(clean_historical["Customer_ID"].duplicated().sum())],
    "Current": [len(clean_current), clean_current.shape[1], int(clean_current.isna().sum().sum()), int(clean_current.duplicated().sum()), int(clean_current["Customer_ID"].duplicated().sum())],
}, index=["Rows", "Columns", "Missing values", "Exact duplicate rows", "Duplicate-ID occurrences"])
display(post_clean_summary)

output_dir = Path("outputs")
output_dir.mkdir(exist_ok=True)
historical_output = output_dir / "Clean_Historical_Customers.csv"
current_output = output_dir / "Clean_Current_Customers.csv"
clean_historical.to_csv(historical_output, index=False)
clean_current.to_csv(current_output, index=False)

print("Exported:", historical_output)
print("Exported:", current_output)
print("All Task 2 post-cleaning checks passed.")

Exported: outputs\Clean_Historical_Customers.csv
Exported: outputs\Clean_Current_Customers.csv
All Task 2 post-cleaning checks passed.


,Historical,Current
Rows,1200,300
Columns,20,19
Missing values,0,0
Exact duplicate rows,0,0
Duplicate-ID occurrences,0,0


## Task 2 Conclusion

Both datasets were assessed before cleaning across all requested quality dimensions. The historical data contained 15 exact repeated records; these were removed, resolving its duplicate IDs without discarding a distinct customer. Category variants were standardized, impossible values were treated as missing, and all missing values were filled using historical reference statistics applied consistently to both datasets.

A final review removed the earlier caps on income, spending, website visits, and support tickets. The reviewed values are statistically unusual but nonnegative and technically possible, and no business maximum or external evidence establishes them as errors. They remain unchanged as legitimate outliers.

The final historical dataset contains **1,200 rows × 20 columns** and the current dataset contains **300 rows × 19 columns**. Both have zero missing values, zero exact duplicates, unique customer IDs, canonical category labels, valid checked domains, and compatible schemas. Timeliness remains unverifiable because no extraction timestamp is supplied. No EDA, encoding, standardization, feature selection, or modelling was performed.

# Task 3 — Descriptive Analytics and Exploratory Data Analysis

Calculate descriptive statistics, create at least five meaningful visualizations, investigate relationships, and write at least five business findings.

*To be completed next.*

# Task 4 — Predictive Model Development and Validation

Build and compare at least two binary classification models using a held-out test set.

*To be completed next.*

# Task 5 — Model Improvement and Final Selection

Apply at least two justified improvement techniques, compare before/after performance, and select the final model.

*To be completed next.*

# Task 6 — Scoring and Prescriptive Retention Plan

Score current customers, calculate financial value and expected benefit, create the retention list, summarize the campaign, and compare top-25 vs top-50 scenarios.

*To be completed next.*

# Task 7 — Final Conclusions and Deliverables

Summarize conclusions and ensure the required files, report, scored dataset, and presentation are ready for submission.

*To be completed next.*